src/classes/RD2_ValidationService_TEST.cls

```diff
@@ -525,7 +525,92 @@ private with sharing class RD2_ValidationService_TEST {
             'The RD should have the household contact as its primary contact');
     }
 
+    /***
+    * @description Verifies DML should fail when installment frequency is updated to null 
+    */
+    @isTest
+    private static void shouldBlockRdUpdateWhenInstallmentFrequencyChangesToNull() {
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        npe03__Recurring_Donation__c rd = getRecurringDonationBuilder()
+            .withContact(getContact().Id)
+            .withInstallmentFrequency(10)
+            .build();
+
+        insert rd;
+
+        rd.InstallmentFrequency__c = null;
+
+        String errMessage = '';
+
+        try {
+            Test.startTest();
+            update rd;
+            Test.stopTest();
+        } catch (DmlException e) {
+            errMessage = e.getMessage();
+        }
+
+        System.assert(errMessage.contains(System.Label.RD2_InstallmentFrequencyMustBeValid),
+            'RD update should fail due to invalid installment frequency with the message: ' + System.Label.RD2_InstallmentFrequencyMustBeValid);
+    }
+
+    /***
+    * @description Verifies DML should fail when installment frequency is updated to a invalid value (greater than twenty) 
+    */
+    @isTest
+    private static void shouldBlockRdUpdateWhenInstallmentFrequencyChangesIsInvalid() {
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        List<npe03__Recurring_Donation__c> rds = getRecurringDonationBuilder()
+            .withContact(getContact().Id)
+            .withInstallmentFrequency(10)
+            .build(2);
+            
+        insert rds;
+
+        rds[0].InstallmentFrequency__c = 21;
+        rds[1].InstallmentFrequency__c = 0;
+
+        String errMessage = '';
+
+        Test.startTest();
+        List<Database.SaveResult> results = Database.update(rds, false);
+        Test.stopTest();
 
+        for (Database.SaveResult result : results) {
+            System.assert(
+                result.getErrors()[0].getMessage().contains(System.Label.RD2_InstallmentFrequencyMustBeValid),
+                'RD update should fail due to invalid installment frequency with the message: ' + System.Label.RD2_InstallmentFrequencyMustBeValid
```

Comment:
- We should also check that `!result.isSuccess()` for failed records and the `result.isSuccess()` for the valid update.
- On the case of error we should display actual error. So it can be `'RD update should fail due to invalid Installment Frequency. Actual message: ' + result.getErrors()[0].getMessage());`